### PR TITLE
Hide phrase delimiter bars outside the talking book tool (BL-11632)

### PR DESCRIPTION
**Pull request by StephenMcConnel**
This change is [Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)

### CODE REVIEW

**Comment by hatton** [comment]
Unit Tests?

**Review by StephenMcConnel — COMMENTED** [review]
I've added a number of unit tests in both C# and typescript.

> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-:-NPK9B0C-dc09wh2DEA3:b-sve5s0)* status: 0 of 11 files reviewed, all discussions resolved

**Review by andrew-polk — CHANGES_REQUESTED** [review]
> Reviewed 11 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)* status: all files reviewed, 7 unresolved discussions (waiting on @StephenMcConnel)

___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/leveledReader/leveledReaderToolboxTool.ts` line 57 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSfF7c9m5pWXOFoSXz:-NPSfF7c9m5pWXOFoSY-:b-3u3fmz) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/leveledReader/leveledReaderToolboxTool.ts#L57)):*
> ```XML
>     }
> 
>     public static restoreAudioSplitMarkers(page: HTMLElement | null) {
> ```

Have you considered storing the dom with \u200B instead of the vertical bar always? Then the leveled reader tool wouldn't need to know about this class/logic, right? Talking book tool could put them back in as well as spreadsheet export.

Or does that introduce backward compatibility issues?
___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/leveledReader/leveledReaderToolboxTool.ts` line 57 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSfmnMAEH2l1ast_8s:-NPSfmnMAEH2l1ast_8t:bw2ryh8) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/leveledReader/leveledReaderToolboxTool.ts#L57)):*
> ```XML
>     }
> 
>     public static restoreAudioSplitMarkers(page: HTMLElement | null) {
> ```

Speaking of backward compatibility, I guess we are going to live with the fact that a book modified in 5.5 which has vertical bars will have those vertical bars hidden in <=5.4 and they won't be able to be unhidden?
___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts` line 136 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSh4TfDbh55CFHR6fA:-NPSh4TfDbh55CFHR6fB:b-uxno8u) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts#L136)):*
> ```XML
>                             fragment.text.substring(leadingClosers.length) +
>                             "</span>";
>                         newHtml += newText;
> ```

I'm not following this logic. Could you include a comment and/or is there a way to make it more straightforward or obvious what is happening?
___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts` line 519 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSguQD0CNIY3k3tztG:-NPSguQD0CNIY3k3tztH:bpqzmte) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts#L519)):*
> ```XML
>     // Remove phrase delimiters used by the talking book tool.
>     const phraseDelimeterRegex = /<span class=["']bloom-audio-split-marker["']>\|/g;
>     textHtml = textHtml.replace(phraseDelimeterRegex, "");
> ```

I don't understand this. How can we remove the opening tag and the content (|) but not the closing tag?
___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts` line 530 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSgbDD3zQfRZF_HIaL:-NPSgbDD3zQfRZF_HIaM:bda2tij) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts#L530)):*
> ```XML
> 
>     // This can sneak in on the current page.
>     const divCogRegex = /<div id="formatButton".*><\/div>/g;
> ```

If I'm not mistaken, these `.*`s want to be `.*?`.
___
*[`src/BloomBrowserUI/bookEdit/toolbox/talkingBook/talkingBook.ts` line 50 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSmiD-0V8uAxXQIO5V:-NPSmiD-0V8uAxXQIO5W:b-ggy9uz) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/talkingBook.ts#L50)):*
> ```XML
>     public async newPageReady(): Promise<void> {
>         this.showImageDescriptionsIfAny();
>         TalkingBookTool.deshroudPhraseDelimiters(ToolBox.getPage());
> ```

I guess it doesn't work to just change the `display` rule while in edit mode to show the vertical bars?
In other words, it would be great to not have to modify the dom back and forth.
___
*[`src/BloomExe/Spreadsheet/SpreadsheetImporter.cs` line 870 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSlkJf4_Rc1W6atpI_:-NPSlkJf4_Rc1W6atpIa:b3d1sac) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomExe/Spreadsheet/SpreadsheetImporter.cs#L870)):*
> ```Smalltalk
> 					var innerContent = editable.InnerXml;
> 					editable.InnerXml = innerContent.Replace("|</span>", "<span class='bloom-audio-split-marker'>|</span></span>");
> 				}
> ```

Shouldn't we make sure we aren't finding a span which already has the `bloom-audio-split-marker class` and wrapping it again?

**Review by StephenMcConnel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)* status: 5 of 13 files reviewed, 7 unresolved discussions (waiting on @andrew-polk)

___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts` line 136 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSh4TfDbh55CFHR6fA:-NPYKKNJA5c7mbC0-sdE:br9p36k) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts#L136)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I'm not following this logic. Could you include a comment and/or is there a way to make it more straightforward or obvious what is happening?
</blockquote></details>

I've added a comment.
___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts` line 519 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSguQD0CNIY3k3tztG:-NPYL1OH9E2fNHqUlkij:b-au6xa6) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts#L519)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I don't understand this. How can we remove the opening tag and the content (|) but not the closing tag?
</blockquote></details>

The </span> is removed by the next Regex expression.  I suppose there's no harm in adding it to this expression.
___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts` line 530 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSgbDD3zQfRZF_HIaL:-NPYKWCz8W141ycMldsp:b-mu15ml) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts#L530)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

If I'm not mistaken, these `.*`s want to be `.*?`.
</blockquote></details>

No, there's an indefinite number of characters to match.  Perhaps the . s want to be [^>] instead.
___
*[`src/BloomBrowserUI/bookEdit/toolbox/talkingBook/talkingBook.ts` line 50 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSmiD-0V8uAxXQIO5V:-NPYIBzJ-xt21vjoXwx5:b-noel7t) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/talkingBook/talkingBook.ts#L50)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

I guess it doesn't work to just change the `display` rule while in edit mode to show the vertical bars?
In other words, it would be great to not have to modify the dom back and forth.
</blockquote></details>

I think changing the dom for the talking book tool is by far the simplest way to achieve the goal of this issue without greatly complicating sentence parsing.
___
*[`src/BloomExe/Spreadsheet/SpreadsheetImporter.cs` line 870 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSlkJf4_Rc1W6atpI_:-NPYISkd4w4xpF5w4p6b:bwzapwz) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomExe/Spreadsheet/SpreadsheetImporter.cs#L870)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Shouldn't we make sure we aren't finding a span which already has the `bloom-audio-split-marker class` and wrapping it again?
</blockquote></details>

The spreadsheet importer won't be creating any spans like that except here.  Those spans are removed on spreadsheet export, replaced by bare | characters.
___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/leveledReader/leveledReaderToolboxTool.ts` line 57 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSfF7c9m5pWXOFoSXz:-NPYIfv4BvxFeJlu80US:b-896fix) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/leveledReader/leveledReaderToolboxTool.ts#L57)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Have you considered storing the dom with \u200B instead of the vertical bar always? Then the leveled reader tool wouldn't need to know about this class/logic, right? Talking book tool could put them back in as well as spreadsheet export.

Or does that introduce backward compatibility issues?
</blockquote></details>

Done.
___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/leveledReader/leveledReaderToolboxTool.ts` line 57 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSfmnMAEH2l1ast_8s:-NPYIgqIA8C2H-t7_ytA:b-l12og4) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/leveledReader/leveledReaderToolboxTool.ts#L57)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Speaking of backward compatibility, I guess we are going to live with the fact that a book modified in 5.5 which has vertical bars will have those vertical bars hidden in <=5.4 and they won't be able to be unhidden?
</blockquote></details>

I've added a hiddenAudioSplitMarkers feature that requires Bloom version 5.5.

**Review by andrew-polk — CHANGES_REQUESTED** [review]
> Reviewed 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)* status: all files reviewed, 1 unresolved discussion (waiting on @StephenMcConnel)

___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts` line 136 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSh4TfDbh55CFHR6fA:-NPYRYTU5tCOjp7GDjYg:b-zcjvte) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts#L136)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary><blockquote>

I've added a comment.
</blockquote></details>

Thanks
___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts` line 530 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSgbDD3zQfRZF_HIaL:-NPYQ4Ao5-r6eDC5xhZN:b-ytybvt) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts#L530)):*
What you have now is just as good, but here's the difference between `*` and `*?` as regex101.com describes them:

> matches the previous token between zero and unlimited times, as many times as possible, giving back as needed (greedy)

vs

> matches the previous token between zero and unlimited times, as few times as possible, expanding as needed (lazy)

So I think that works as well.



_Code quote:_
```XML
<div id="formatButton".*><img.*><\/div>
```
___
*[`src/BloomExe/Spreadsheet/SpreadsheetImporter.cs` line 870 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSlkJf4_Rc1W6atpI_:-NPYRgRc-tR0QPNyzBhi:b416uvr) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomExe/Spreadsheet/SpreadsheetImporter.cs#L870)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary><blockquote>

The spreadsheet importer won't be creating any spans like that except here.  Those spans are removed on spreadsheet export, replaced by bare | characters.
</blockquote></details>

Maybe I was thinking too pathologically... but I'm thinking of someone putting that markup in themselves (likely copied from some other book).

Anyway, I think it is moot now that the exporter only puts out the `|` without the span.
But based on that, I think this code needs to be updated to drop the span. (And maybe some new unit test needs to be added?)

**Review by StephenMcConnel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)* status: all files reviewed, 1 unresolved discussion (waiting on @andrew-polk)

___
*[`src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts` line 530 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSgbDD3zQfRZF_HIaL:-NPYX8Ty4U8AiE0sasql:b-k5xhpu) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomBrowserUI/bookEdit/toolbox/readers/libSynphony/jquery.text-markup.ts#L530)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

What you have now is just as good, but here's the difference between `*` and `*?` as regex101.com describes them:

> matches the previous token between zero and unlimited times, as many times as possible, giving back as needed (greedy)

vs

> matches the previous token between zero and unlimited times, as few times as possible, expanding as needed (lazy)

So I think that works as well.
</blockquote></details>

OK, I forget that some variants of regex have the non-greedy matching.
___
*[`src/BloomExe/Spreadsheet/SpreadsheetImporter.cs` line 870 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSlkJf4_Rc1W6atpI_:-NPYXT_32nQxpegeWpyx:b-qdu3b3) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomExe/Spreadsheet/SpreadsheetImporter.cs#L870)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Maybe I was thinking too pathologically... but I'm thinking of someone putting that markup in themselves (likely copied from some other book).

Anyway, I think it is moot now that the exporter only puts out the `|` without the span.
But based on that, I think this code needs to be updated to drop the span. (And maybe some new unit test needs to be added?)
</blockquote></details>

OK, what I think you're saying is that the importer should
1) drop any spans like <span class="bloom-audio-split-marker">\u200B</span> from the InnerXml
before
2) insert the span to replace the | at the end of a span (ie, the current behavior)

**Review by andrew-polk — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)* status: all files reviewed, 1 unresolved discussion (waiting on @StephenMcConnel)

___
*[`src/BloomExe/Spreadsheet/SpreadsheetImporter.cs` line 870 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSlkJf4_Rc1W6atpI_:-NPc3-5qCtOI0zbiH48v:bmaq06j) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomExe/Spreadsheet/SpreadsheetImporter.cs#L870)):*
Sorry; I was unclear.
By 
>  I think this code needs to be updated to drop the span

I meant I think 
`editable.InnerXml = innerContent.Replace("|</span>", "<span class='bloom-audio-split-marker'>\u200B</span></span>");`
should now be 
`editable.InnerXml = innerContent.Replace("|", "<span class='bloom-audio-split-marker'>\u200B</span>");`

I'm not sure about the two lines you added to take out any `bloom-audio-split-marker` spans. Seems like if they are there, we probably want to leave them in? What I was worried about before was creating an invalid double-nested span. But we can't do that now, right? So I think it would probably be best to just take those two lines back out. Sorry for the confusion.

**Review by StephenMcConnel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)* status: 11 of 13 files reviewed, 1 unresolved discussion (waiting on @andrew-polk)

___
*[`src/BloomExe/Spreadsheet/SpreadsheetImporter.cs` line 870 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSlkJf4_Rc1W6atpI_:-NPrpByO1fj4nKag0_lj:b-896fix) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomExe/Spreadsheet/SpreadsheetImporter.cs#L870)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Sorry; I was unclear.
By 
>  I think this code needs to be updated to drop the span

I meant I think 
`editable.InnerXml = innerContent.Replace("|</span>", "<span class='bloom-audio-split-marker'>\u200B</span></span>");`
should now be 
`editable.InnerXml = innerContent.Replace("|", "<span class='bloom-audio-split-marker'>\u200B</span>");`

I'm not sure about the two lines you added to take out any `bloom-audio-split-marker` spans. Seems like if they are there, we probably want to leave them in? What I was worried about before was creating an invalid double-nested span. But we can't do that now, right? So I think it would probably be best to just take those two lines back out. Sorry for the confusion.
</blockquote></details>

Done.

**Review by andrew-polk — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)* status: all files reviewed, 1 unresolved discussion (waiting on @StephenMcConnel)

___
*[`src/BloomExe/Spreadsheet/SpreadsheetImporter.cs` line 870 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSlkJf4_Rc1W6atpI_:-NPrt_uY5UW4xzgUVkOD:bblf9vu) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomExe/Spreadsheet/SpreadsheetImporter.cs#L870)):*
Thanks.

Do you agree with
> I meant I think
`editable.InnerXml = innerContent.Replace("|</span>", "<span class='bloom-audio-split-marker'>\u200B</span></span>");`
should now be
`editable.InnerXml = innerContent.Replace("|", "<span class='bloom-audio-split-marker'>\u200B</span>");`

?

**Review by StephenMcConnel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)* status: all files reviewed, 1 unresolved discussion (waiting on @andrew-polk)

___
*[`src/BloomExe/Spreadsheet/SpreadsheetImporter.cs` line 870 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSlkJf4_Rc1W6atpI_:-NPsjzqx6Dt0xJh2pQBS:b-q150f3) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomExe/Spreadsheet/SpreadsheetImporter.cs#L870)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary><blockquote>

Thanks.

Do you agree with
> I meant I think
`editable.InnerXml = innerContent.Replace("|</span>", "<span class='bloom-audio-split-marker'>\u200B</span></span>");`
should now be
`editable.InnerXml = innerContent.Replace("|", "<span class='bloom-audio-split-marker'>\u200B</span>");`

?
</blockquote></details>

My thinking was that the audio processing would always end a span with the phrase marker ("|") if there was audio in the book.  If the book didn't actually contain audio, then any existing "|" characters would be left alone since they may have been intended to be in the text verbatim.  But if you think that's never going to happen, I can certainly simplify the match and replace.

**Review by andrew-polk — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)* status: all files reviewed, 1 unresolved discussion (waiting on @StephenMcConnel)

___
*[`src/BloomExe/Spreadsheet/SpreadsheetImporter.cs` line 870 at r1](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663#-NPSlkJf4_Rc1W6atpI_:-NPtExIN3hdRnjRQp_e6:b-70a2y8) ([raw file](https://github.com/BloomBooks/BloomDesktop/blob/d0b4121fae864f4dbbd5b2cca740bc99654faba1/src/BloomExe/Spreadsheet/SpreadsheetImporter.cs#L870)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary><blockquote>

My thinking was that the audio processing would always end a span with the phrase marker ("|") if there was audio in the book.  If the book didn't actually contain audio, then any existing "|" characters would be left alone since they may have been intended to be in the text verbatim.  But if you think that's never going to happen, I can certainly simplify the match and replace.
</blockquote></details>

Ok, so you're trying to distinguish between vertical bars which are there for audio and those which aren't? And trying not to change the ones which aren't? Got it.

Now that I understand that, I'm thinking of the scenario where a user is inputting new text via spreadsheet and wants to add vertical bars for doing audio splits.
So now we're balancing the need of a user who wants to insert a vertical bar and not have it mean audio split vs the one who wants to add a vertical bar for audio without having to include the html...
If they open the talking book tool, it will change the dom to have the span and class, right? So maybe it is best to just mimic that behavior? If they don't want to record audio, having that span won't hurt anything, will it?
As a side note, I like the idea of the import and export being reciprocal logic (one replace A with B; the other replaces B with A).

What do you think?

**Review by andrew-polk — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/BloomBooks/BloomDesktop/5663)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @StephenMcConnel)
